Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2021, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PolygonalSelectTool.h"
+#include "../ImageEditor.h"
+#include "../Layer.h"
+#include <AK/Queue.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/Button.h>
+#include <LibGUI/ComboBox.h>
+#include <LibGUI/ItemListModel.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Model.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/ValueSlider.h>
+
+namespace PixelPaint {
+
+void PolygonalSelectTool::flood_polygon_selection(Gfx::Bitmap& polygon_bitmap, Gfx::IntPoint polygon_delta)
+{
+    VERIFY(polygon_bitmap.bpp() == 32);
+
+    // Create Mask which will track already-processed pixels
+    Mask flood_mask = Mask::empty(polygon_bitmap.rect());
+    Mask selection_mask = Mask::full(polygon_bitmap.rect().translated(polygon_delta));
+    Queue<Gfx::IntPoint> points_to_visit = Queue<Gfx::IntPoint>();
+
+    // flood fill outside the polygon, clearing the selection mask for any pixel we can visit
+    points_to_visit.enqueue({ polygon_bitmap.width() - 1, polygon_bitmap.height() - 1 });
+    selection_mask.set(Gfx::IntPoint(polygon_bitmap.width() - 1, polygon_bitmap.height() - 1).translated(polygon_delta), 0);
+    flood_mask.set(polygon_bitmap.width() - 1, polygon_bitmap.height() - 1, 0);
+
+    while (!points_to_visit.is_empty()) {
+        auto current_point = points_to_visit.dequeue();
+        auto candidate_points = Array {
+            current_point.moved_left(1),
+            current_point.moved_right(1),
+            current_point.moved_up(1),
+            current_point.moved_down(1)
+        };
+        for (auto candidate_point : candidate_points) {
+            if (!polygon_bitmap.rect().contains(candidate_point)) {
+                continue;
+            }
+            if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && polygon_bitmap.get_pixel(candidate_point.x(), candidate_point.y()) != Color::Black) {
+
+                polygon_bitmap.set_pixel(candidate_point.x(), candidate_point.y(), Color::Black);
+                points_to_visit.enqueue(candidate_point);
+                selection_mask.set(Gfx::IntPoint(candidate_point.x(), candidate_point.y()).translated(polygon_delta), 0);
+            }
+            flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+        }
+    }
+
```

Comment:
This flood fill is very similar to the ones used in your other tools, could it be factored out into a shared function?

Reply:
The honest answer here is that I don't know. It feels like it should though. Just off top of my head, the key difference here is what happens when you reach a pixel. In the bucket tool you color the pixel, in the Polygon Select Tool you toggle a point in a mask.

Maybe some abstracted "flood fill" class/method where you pass it a lambda of what to do for each pixel it reaches? I'm not sure where a good place to put such a thing would be. What about adding flood_fill to Bitmap itself in LibGFX and accepting a lambda for the action to perform on each pixel reached?

Reply:
@MacDue I have an implementation of this locally and it seems to work just fine. It cleans up the tool code *a bunch*. I may extract this change into a separate branch and create a separate PR for it, and then rebase this (and another selection tool I'm working on) once that lands, or some variation of it :).

The abstraction approach can be discussed in that PR but to give an idea of what I currently have working, here's a couple snippets.

I currently have a function in Bitmap
```
void flood_visit_from_point(Gfx::IntPoint const& start_point, int threshold,Optional<Color> target_color, Function<void(Gfx::IntPoint location)> pixel_reached);
```
And then in a tool like Bucket Tool:

```
auto location_reached = [&](Gfx::IntPoint location) {
    bitmap.set_pixel<Gfx::StorageFormat::BGRA8888>(location.x(), location.y(), fill_color);
};

bitmap.flood_visit_from_point(start_position, threshold,target_color ,move(location_reached));
```

Reply:
Marking this resolved since the extraction of flood fill logic has landed in master and most recent push of this PR leverages the new code :)